### DISCUSSION TITLE
How use npm install --build-from-source in pnpm

### DISCUSSION THREAD

**Discussion by v9u** [pnpm/pnpm#5665] [Q&A]
How use npm install --build-from-source in pnpm?
I need use sqlite3 in Electron,like this:
`
npm install sqlite3 --build-from-source --runtime=electron --target=18.2.1 --dist-url=https://electronjs.org/headers
`
but  ` pnpm install sqlite3 -- --build-from-source --runtime=electron --target=18.2.1 --dist-url=https://electronjs.org/headers` is not work
> [Electron Sqlite3 Install Details](https://github.com/TryGhost/node-sqlite3#custom-builds-and-electron)
>pnpm : `7.17.0`
>node : `v16.15.1`
>electron: `v19.07`

**Comment by polnikale** [comment]
any answer?

**Reply by dgtllion** [reply]
I'm facing the same problem. Is there any update? 2 years later...

**Comment by phiresky** [comment]
@polnikale @dgtllion 

I have found the simplest solution to be to use `pnpm patch <package-name>` to directly patch the arguments you want into into the dependency (wherever appropriate). For example, arguments you could directly add to the dependency's package.json. file.

It's not beautiful, but that way you guarantee that the arguments are passed to the build script directly when running `pnpm install`, no other steps needed!

**Comment by Galadirith** [comment]
Since this discussion is one of the top Google results when searching for installing packages from source using `pnpm`, I wanted to share my related issue and eventual solution. Noting that my issue is only related to the title issue of `--build-from-source`. Unfortunately I do not have a solution to the other additional CLI options, and specifically targeting an `electron` runtime.

I had a problem with installing `sqlite3` on an Apple Silicon Mac. Specifically I was unable to use `pnpm` to install n8n because it was silently failing to correctly install the `sqlite3` dependency because there are no prebuild binaries for Apple Silicon because of the resolution issues with the `arm` `{arch}` as detailed in the [node-sqlite3 README](https://github.com/TryGhost/node-sqlite3?tab=readme-ov-file#prebuilt-binaries).

In the end my solution was to install a tagged version from GitHub (https://pnpm.io/cli/add)

```
pnpm add github:TryGhost/node-sqlite3#v5.1.7
```

This forces a build from source because you are adding a source dependency. You will likely be prompted to approve a build from source

```
╭ Warning ───────────────────────────────────────────────────────────────────────────────────╮
│                                                                                            │
│   Ignored build scripts: sqlite3.                                                          │
│   Run "pnpm approve-builds" to pick which dependencies should be allowed to run scripts.   │
│                                                                                            │
╰────────────────────────────────────────────────────────────────────────────────────────────╯
```

which you can follow and do

```
pnpm approve-builds
```

which will then prompt you with `sqlite3` as a build that needs approval, which you can authorise, and then done. You might need to rerun the `pnpm add github...` command, I am not sure if it is necessary, I spent the extra 10s to rerun just incase it was needed. Then finally

```
pnpx n8n
```

now works. This of course means you can't use nicer features like upgrading packages, you would have to manually check and modify the github deps. In my case that is perfectly acceptable, but understandable if that isn't for others.

I do not have any recent experience working with electron, and certainly none attempting to use `pnpm` to manage dependencies for an electron project. But for whatever it is worth my own reflection is that `pnpm` is really valuable when you need to manage dependencies for node runtime projects (where you might have 10+ projects all referencing a similar set of dependencies), but less useful when you want to manage dependencies for electron runtime projects (where you might only have 1-2 electron runtime projects). I think [Why You Should Use npm (and Not pnpm... Yet) to Build Electron + React + Vite + Tailwind Apps](https://dev.to/yurirxmos/why-you-should-use-npm-and-not-pnpm-yet-to-build-electron-react-vite-tailwind-apps-4oc9) summarises the problems you might encounter. `pnpm` seems very much built to replace `npm` for `node` runtimes, so incompatibilities seems likely, even before trying to deal with building from source. I am sure there are good reasons why it is valuable using `pnpm` for electron apps that I am not aware of, but maybe it's worth considering if you want to spend the time and deal with potential hacky solutions to make `pnpm` work with dep management for an electron app, over `npm` that might not have the additional desirable features of `pnpm` but `npm` is (more) likely to just work, and spend the saved time building some great electron apps 😌
